### ISSUE TITLE
Memory Leak

### ISSUE THREAD

**Issue by agracia-foticos** [unjs/unhead#281]
### Environment

- Operating System: `Linux`
- Node Version:     `v20.5.1`
- Nuxt Version:     `3.8.2`
- CLI Version:      `3.10.0`
- Nitro Version:    `2.8.0`
- Package Manager:  `npm@9.8.0`
- Builder:          `-`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`

### Reproduction

![image](https://github.com/unjs/unhead/assets/77386602/80296784-0220-41e2-8277-db4498df9959)

When i crawled our web, crashes by memory leak

### Describe the bug

![image](https://github.com/unjs/unhead/assets/77386602/80296784-0220-41e2-8277-db4498df9959)

When i crawled our web, crashes by memory leak

**Comment by harlan-zw** [comment]
Will need some reproduction details

**Comment by agracia-foticos** [comment]
> Will need some reproduction details

We have a lot of code generated, it is impossible to have a minimal playback environment

**Comment by harlan-zw** [comment]
Without a reproduction, there isn't much I can do. 

There's no way to confirm if Unhead is the problem or if downstream code implementing it, which seems more likely given this is the first time a memory leak issue has been opened and knowing how the API is designed.

**Comment by BobbieGoede** [comment]
While testing for memory leaks in `@nuxtjs/i18n` it did seem like something was leaky and I suspected `unhead`, but I haven't been able to consistently reproduce it. 

Whatever seemed to be leaking did so at a low rate, unless someone manages to reproduce it, I wouldn't worry too much about it. 😅

**Comment by harlan-zw** [comment]
Thank you for your investigation @BobbieGoede! Some great work on @nuxtjs/i18n :clap:

**Comment by dargmuesli** [comment]
I investigated too :sob: :joy:  I have one or two other theories on suspect code, maybe we can have a look together in the coming days @harlan-zw or @BobbieGoede? If you have time next week, we could meet on Discord to stop the leak :broom:

**Comment by dargmuesli** [comment]
Just saw and tested that https://github.com/nuxt-modules/i18n/pull/2621 is also a great fix :fire: :rocket:  From the looks it seems that the only directly noticeable leak left is on me in my project. So no need for a call rn, but I may get back to that if I find anything else. Thank you all for the good work!

**Comment by MosheL** [comment]
I can explain better than the opener of the bug. He is right.
I was wasted a day to debug a similar bug in my code, so I can explain better.

The leak is in the `$vm._` of vue,  persisted in node memory by the global `__unhead_injection_handler__` and its friends, so after the SSR page go down, the memory is not freed and stayed in memory again and again.


![image](https://github.com/unjs/unhead/assets/2265043/6dd10dce-9919-48be-ad0b-a1c5ff5a901b)

**Comment by harlan-zw** [comment]
There should only be a single instance of Vue kept in memory, it's not freed as you correctly identified.

It shouldn't be stacking though, your screenshot doesn't show that it is afaik?

**Comment by BobbieGoede** [comment]
BTW, while I was searching for what could be leaking memory in the i18n module I often saw [this pattern](https://github.com/search?q=repo%3Avuejs%2Fcore+%2FSymbol%5C%28__DEV__+%5C%3F%2F&type=code) (`Symbol(__DEV__ ? 'LIB_KEY_HERE' : '')` which obscures/removes the identifying global key in a production build. 

I guess the primary reason for doing this is to reduce the build size, but a side effect is that we don't see these keys when debugging memory (which I have only done on a production build). Since unhead does not use this pattern, and it uses some memory, it will be one of the first things you'll see when debugging which doesn't necessarily mean that it is leaking.

**Comment by obulat** [comment]
This seems to be one of the largest memory leaks that's preventing us at [Openverse](https://openverse.org) from migrating to Nuxt 3. Even after I removed all of the `useHead` calls, only leaving the `head` properties in the `nuxt.config.ts`, I can see a lot of retained dependencies of `$head`:

<img width="1230" alt="Screenshot 2024-01-25 at 6 08 59 PM" src="https://github.com/unjs/unhead/assets/15233243/2d05a90e-edf9-47c0-b73e-2908ef812ad6">

Here's the [PR for migrating to Nuxt 3](https://github.com/WordPress/openverse/pull/3646) that is, unfortunately, unusable, due to memory leaks.

**Comment by harlan-zw** [comment]
Looks like a vue-router stack? $head is a singleton on the vue VM (like globalProperties), so not sure what you're showing here exactly.

Not to say there isn't an issue, but this doesn't help me identify anything from Unhead's side.

**Comment by harlan-zw** [comment]
I did some testing with [clinic.js](https://clinicjs.org/) and was not able to replicate any sort of memory leak.

![image](https://github.com/unjs/unhead/assets/5326365/9e9383b7-c27a-4778-baa6-04f3139d1f95)

This is with ~500 connections doing ~5k requests x2.

Will leave this issue open in case anyone has any further research they can share. (I tried testing with other clinic.js tools but it was failing).

**Comment by Rigo-m** [comment]
![image](https://github.com/user-attachments/assets/f5a797cc-00b4-411e-a6c3-854a3313e7ca)
I am too having problems with `__unhead_injection_handler__` retaining a lot of stuff.

**Comment by Rigo-m** [comment]
Also, i see strings not garbage collected
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/696951ee-b9b8-4a62-a03f-e26e35640fab">
@harlan-zw up for a call if you want to help me check this out

**Comment by daniluk4000** [comment]
> Also I see many strings not getting garbage collected, megs and megs of ram not being cleared. Up for a call if you want to check this out @harlan-zw 

Can you include reproduction? On our large project with heavy unhead usage (including not-so-official solutions in our plugins) our memory stays at ~400MB without significant increases. I also noticed unhead growing in size, but GC seems to be working for us

**Comment by Tofandel** [comment]
To anyone experiencing this issue, I would ask that you check in your package-lock.json or yarn.lock that your vue is up to date, (it might not be because of a bug in yarn that does not update sub deps when targeting update of a single package so if only nuxt is a direct dep and is targeted for upgrade, vue will not upgrade)


Please check that you are at least on 3.5.7 which fixes 3 memory leak issues, and if you are not, then please try again after upgrading and report back
https://github.com/vuejs/core/blob/main/CHANGELOG.md#357-2024-09-20

**Comment by Xenossolitarius** [comment]
Managed to trigger the memory leak and 100% concisely.  I am doing some workaround here in Nuxt to dodge Nuxt dedupe bcs my object is quite quite large. For the info it was working for some time and didn't cause any memory leaks but i buffed the deps to latest and now it's exploding. I am using huge payloads od 300kb so as you know the server fails in seconds.

```
export const useBigDataFetch = <T extends JSONValue>(
  path: string | (() => string),
  _args?: UseFetchOptions<T>,
) => {
  const url = typeof path === 'string' ? path : path()
  const key = url.replaceAll('/', '_')

  const rawData = useState(key, () => shallowRef(new JSONPayload<T>(key))) // shallow by default

  const error = ref<unknown>()

  const { locale } = useFetchOptions()

  const execute = async (force?: boolean) => {
    if (!force && rawData.value.payload) return
    try {
      const nuxt = useNuxtApp()

      const res = await $casinoInternalFetch<T>(url)
      const data = new JSONPayload<T>(key, res)

      nuxt.runWithContext(() => {
        // embed in html directly
        useServerHead({
          script: [
            {
              tagPosition: 'bodyClose', // near rest of nuxt data
              innerHTML: `window.value.key =` + data.raw,
            },
          ],
        })
      })

      rawData.value = data
    } catch (e) {
      error.value = e
    }
  }

  const data = computed(() => rawData.value.payload)

  watch(locale, () => execute(true))

  return {
    data,
    error,
    execute,
  }
}
```

I can workaround around this issue easily but still this is the common use case of How not to use useHead that works from version to version. 

Just ping me if you need any more data on the issue but schema, nuxt, vue, unhead are used in unhoisted pnpm monorepo. 

PS. as a user I would expect the default behavior of what I am doing here to "not work" in non failing way: "produce a lot of undefineds" but I would not like to produce a memory leak.

**Comment by cjpearson** [comment]
@Xenossolitarius Does it make any difference if you move `const nuxt = useNuxtApp()` outside the execute function? Generally composables should only be called from the main body of a composable or setup function.

**Comment by Xenossolitarius** [comment]
Not really. I moved it all over the place but this is not supported anymore. Its positioned after the primary await in app.vue

**Comment by harlan-zw** [comment]
This should be solved in v2 which is now in alpha. There's a migration guide here: https://unhead-unjs-io.nuxt.dev/docs/migration
